### PR TITLE
Improvements for amoc

### PR DESCRIPTION
**Pull request by michalwski**
Various changes made when using escalus with Amoc.

Supersedes  #59

### CODE REVIEW

File: src/escalus_stanza.erl [hunk 1]
```diff
@@ -639,22 +640,50 @@ resume(SMID, PrevH) ->
...
           attrs = [mam_ns_attr(), {<<"queryid">>, QueryId}],
-          children = defined(Children)}]).
+          children = ChildEl}]).
+
 
 mam_lookup_messages_iq(QueryId, Start, End, WithJID) ->
-    mam_archive_query(QueryId, [fmapM(fun start_elem/1, Start),
-                                fmapM(fun end_elem/1, End),
-                                fmapM(fun with_elem/1, WithJID)]).
+    mam_archive_query(QueryId, [field_el(<<"FORM_TYPE">>, <<"hidden">>, ?NS_MAM),
```
Comment (gadgetci) [inline src/escalus_stanza.erl]: According to **Elvis**:
> Line 682 is too long:     mam_archive_query(QueryId, [field_el(<<"FORM_TYPE">>, <<"hidden">>, ?NS_MAM),.

File: src/escalus_pred.erl [hunk 2]
```diff
@@ -273,7 +283,7 @@ is_0184_receipt(Request, ProperResFrom,
             is_0184_receipt(Request, ProperResFrom,
                             Receipt#xmlel{ children = [Received] })
     end;
-is_0184_receipt(_, _, _) ->
+is_0184_receipt(_,_,_) ->
```
Comment (gadgetci) [inline src/escalus_pred.erl]: According to **Elvis**:
> Missing space after "," on line 286

File: src/escalus_pred.erl [hunk 3]
```diff
@@ -212,6 +215,13 @@ is_mam_archived_message(Msg, #xmlel{} = Stanza) ->
                                  {element, <<"message">>}]),
     is_chat_message(Msg,M).
```
Comment (gadgetci) [inline src/escalus_pred.erl]: According to **Elvis**:
> Missing space after "," on line 216

**Comment by michalwski** [comment]
This PR was tried here: michalwski/MongooseIM#15

**Comment by michalwski** [comment]
@elbrujohalcon, any idea why gadget/compiler and gadget/xref failed? The reason here is truncated so it's hard to guess the root cause.

**Comment by elbrujohalcon** [comment]
@michalwski I'll check

**Comment by elbrujohalcon** [comment]
```
fatal error: expat.h: No such file or directory
```

We've dealt with this before. Probably @demian711 or @ignaciom can help fixing it again.

File: src/escalus_pred.erl [hunk 4]
```diff
@@ -468,6 +478,21 @@ has_feature(Feature, Stanza) ->
...
 
+has_field_with_type(Name, Type, Form) ->
+    Fields = exml_query:paths(Form, [{element, <<"field">>}]),
+    lists:any(fun(Item) ->
+                      exml_query:attr(Item, <<"var">>) == Name
+                      andalso
+                      exml_query:attr(Item, <<"type">>) == Type
+              end, Fields).
+
+has_field_value(Name, Value, Form) ->
```
Comment (gadgetci) [inline src/escalus_pred.erl]: According to **Compiler**:
> Warning: missing specification for function has_field_value/3

File: src/escalus_stanza.erl [hunk 5]
```diff
@@ -639,22 +640,51 @@ resume(SMID, PrevH) ->
...
+                                      children = [#xmlcdata{content = E}]}
+                       end, Values),
+    #xmlel{name = <<"field">>,
+           attrs = [{<<"type">>, Type},
+                    {<<"var">>, Name}],
+           children = Fields };
+field_el(Name, Type, Value) ->
+    field_el(Name, Type, [Value]).
 
 mam_archive_query(QueryId) ->
```
Comment (gadgetci) [inline src/escalus_stanza.erl]: According to **Compiler**:
> Warning: missing specification for function mam_archive_query/1

File: src/escalus_stanza.erl [hunk 6]
```diff
@@ -639,22 +640,51 @@ resume(SMID, PrevH) ->
...
-      <<"get">>,
+      <<"set">>,
       [#xmlel{
           name = <<"query">>,
           attrs = [mam_ns_attr(), {<<"queryid">>, QueryId}],
-          children = defined(Children)}]).
+          children = ChildEl}]).
+
 
 mam_lookup_messages_iq(QueryId, Start, End, WithJID) ->
```
Comment (gadgetci) [inline src/escalus_stanza.erl]: According to **Compiler**:
> Warning: missing specification for function mam_lookup_messages_iq/4

File: src/escalus_stanza.erl [hunk 7]
```diff
@@ -639,22 +640,51 @@ resume(SMID, PrevH) ->
...
-                                fmapM(fun with_elem/1, WithJID)]).
+    Fields = [field_el(<<"FORM_TYPE">>, <<"hidden">>, ?NS_MAM),
+              field_el(<<"start">>, <<"text-single">>, Start),
+              field_el(<<"end">>, <<"text-single">>, End),
+              field_el(<<"with">>, <<"jid-single">>, WithJID)
+             ],
+    mam_archive_query(QueryId, [Fields]).
 
 %% Include an rsm id for a particular message.
 mam_lookup_messages_iq(QueryId, Start, End, WithJID, DirectionWMessageId) ->
```
Comment (gadgetci) [inline src/escalus_stanza.erl]: According to **Compiler**:
> Warning: missing specification for function mam_lookup_messages_iq/5

File: src/escalus_stanza.erl [hunk 8]
```diff
@@ -639,22 +640,51 @@ resume(SMID, PrevH) ->
 %%
 %% @TODO: move the stanza constructors from
 %% tests/mam_SUITE.erl into here.
+field_el(_Name, _Type, undefined) ->
```
Comment (gadgetci) [inline src/escalus_stanza.erl]: According to **Compiler**:
> Warning: missing specification for function field_el/3

File: src/escalus_pred.erl [hunk 9]
```diff
@@ -468,6 +478,21 @@ has_feature(Feature, Stanza) ->
               end,
               Features).
 
+has_field_with_type(Name, Type, Form) ->
```
Comment (gadgetci) [inline src/escalus_pred.erl]: According to **Compiler**:
> Warning: missing specification for function has_field_with_type/3

File: src/escalus_pred.erl [hunk 10]
```diff
@@ -210,8 +213,15 @@ is_mam_archived_message(Msg, #xmlel{} = Stanza) ->
     M = exml_query:path(Stanza, [{element, <<"result">>},
                                  {element, <<"forwarded">>},
                                  {element, <<"message">>}]),
-    is_chat_message(Msg,M).
+    is_chat_message(Msg, M).
 
+is_mam_fin_message(Stanza) ->
```
Comment (gadgetci) [inline src/escalus_pred.erl]: According to **Compiler**:
> Warning: missing specification for function is_mam_fin_message/1

File: src/escalus_pred.erl (hunk 4 above)
Comment (gadgetci) [inline src/escalus_pred.erl]: According to **Compiler**:
> Warning: variable 'Name' is unused

**Comment by elbrujohalcon** [comment]
@michalwski gadget fixed… now you need to fix your code ;)

**Comment by michalwski** [comment]
Great, I can do that now :) Thanks a lot :)

File: src/escalus_stanza.erl [hunk 11]
```diff
@@ -639,24 +640,60 @@ resume(SMID, PrevH) ->
...
+              field_el(<<"end">>, <<"text-single">>, End),
+              field_el(<<"with">>, <<"jid-single">>, WithJID)
+             ],
+    mam_archive_query(QueryId, [Fields]).
 
 %% Include an rsm id for a particular message.
+-spec mam_lookup_messages_iq(binary(), binary(), binary(), binary(), term()) ->
+    exml:element().
 mam_lookup_messages_iq(QueryId, Start, End, WithJID, DirectionWMessageId) ->
     IQ = #xmlel{children=[Q]} = mam_lookup_messages_iq(QueryId, Start, End, WithJID),
```
Comment (gadgetci) [inline src/escalus_stanza.erl]: According to **Elvis**:
> Line 698 is too long:     IQ = #xmlel{children=[Q]} = mam_lookup_messages_iq(QueryId, Start, End, WithJID),.

**Comment by michalwski** [comment]
Great, @gadgetci is now satisfied with my pull request!

**Pull request merged**